src/uwtools/drivers/upp.py

```diff
@@ -0,0 +1,163 @@
+        batch: bool = False,
+    ):
+        """
+        The driver.
+
+        :param cycle: The cycle.
+        :param leadtime: The leadtime.
+        :param config: Path to config file (read stdin if missing or None).
+        :param dry_run: Run in dry-run mode?
+        :param batch: Run component via the batch system?
+        """
+        super().__init__(
+            config=config,
+            dry_run=dry_run,
+            batch=batch,
+            cycle=cycle,
+            leadtime=leadtime,
+        )
+        if self._dry_run:
+            dryrun()
+        self._cycle = cycle
+        self._leadtime = leadtime
+
+    # Workflow tasks
+
+    @tasks
+    def files_copied(self):
+        """
+        Files copied for run.
+        """
+        yield self._taskname("files copied")
+        yield [
+            filecopy(src=Path(src), dst=self._rundir / dst)
+            for dst, src in self._driver_config.get("files_to_copy", {}).items()
+        ]
+
+    @tasks
+    def files_linked(self):
+        """
+        Files linked for run.
+        """
+        yield self._taskname("files linked")
+        yield [
+            symlink(target=Path(target), linkname=self._rundir / linkname)
+            for linkname, target in self._driver_config.get("files_to_link", {}).items()
+        ]
+
+    @task
+    def namelist_file(self):
+        """
+        The namelist file.
+        """
+        path = self._namelist_path
+        yield self._taskname(str(path))
+        yield asset(path, path.is_file)
+        yield None
+        path.parent.mkdir(parents=True, exist_ok=True)
+        self._create_user_updated_config(
+            config_class=NMLConfig,
+            config_values=self._driver_config["namelist"],
+            path=path,
+        )
+
+    @tasks
+    def provisioned_run_directory(self):
+        """
+        Run directory provisioned with all required content.
+        """
+        yield self._taskname("provisioned run directory")
+        yield [
+            self.files_copied(),
+            self.files_linked(),
+            self.namelist_file(),
+            self.runscript(),
+        ]
+
+    @task
+    def runscript(self):
+        """
+        The runscript.
+        """
+        path = self._runscript_path
+        yield self._taskname(path.name)
+        yield asset(path, path.is_file)
+        yield None
+        self._write_runscript(path=path, envvars={})
+
+    # Private helper methods
+
+    @property
+    def _driver_name(self) -> str:
+        """
+        Returns the name of this driver.
+        """
+        return STR.upp
+
+    @property
+    def _namelist_path(self) -> Path:
+        """
+        Path to the namelist file.
+        """
+        return self._rundir / "itag"
+
+    @property
+    def _runscript_path(self) -> Path:
+        """
+        Path to the runscript.
+        """
+        return self._rundir / f"runscript.{self._driver_name}"
+
+    @property
+    def _runcmd(self) -> str:
```

Comment:
It just depends on how the driver needs to be run. Historically, a lot of NWP components operate just by running the program (possibly preceded by `mpirun`, `mpiexec`, or `srun` if they are MPI parallel programs), with no arguments, because historically Fortran did not support command-line arguments. So, for a serial program, you'd just run
```
./component
```
or for a parallel MPI program, to run it with 8 parallel tasks
```
mpirun -n 8 component
```
It was assumed that any configuration required by the program would be supplied by namelists in a namelist file, and the program would open a file with a standard name that the user couldn't change (like the `itag` for `upp` as seen in this PR). Funny enough, MPI is newer than Fortran and so it _does_ support command like arguments, but those just tell MPI how to behave, and the program still gets its configuration from a namelist file.

C programs have always supported command like arguments, and modern Fortran programs do, too, so some components we're writing drivers for need those. So our default `_runcmd()` method in the `Driver` base class supports the classic Fortran-style no-argument invocation, and drivers that need extra command-line arguments need to override that method and provide their own implementation.

Yours is a good question, though: Maybe we should make the `Driver` base version of `_runcmd()` more flexible so that it's less likely that we'll need to override it. You might consider brining it up in standup and potentially writing a ticket for the work if the team likes the idea.

Reply:
Thats a good point, and the fewer overrides the better